optuna/samplers/_nsga2.py

```diff
@@ -327,3 +395,52 @@ def _crowding_distance_sort(population: List[FrozenTrial]) -> None:
 
     population.sort(key=lambda x: manhattan_distances[x.number])
     population.reverse()
+
+
+def _constrained_dominates(
+    trial0: FrozenTrial, trial1: FrozenTrial, directions: Sequence[StudyDirection]
+) -> bool:
+    """Checks constrained-domination.
+
+    A trial x is said to constrained-dominate a trial y, if any of the following conditions is
+    true:
+    1) Trial x is feasible and trial y is not.
+    2) Trial x and y are both infeasible, but solution x has a smaller overall constraint
+    violation.
+    3) Trial x and y are feasible and trial x dominates trial y.
+    """
+
+    constraints0 = trial0.system_attrs[_CONSTRAINTS_KEY]
+    constraints1 = trial1.system_attrs[_CONSTRAINTS_KEY]
```

Comment:
Could this dictionary lookup ever fail, for instance when a user would resume a once finished study? If so, we could raise a more user friendly error here. Could be a follow up.

Reply:
Thank you for pointing it out. I think we can exclude the trials without constraint values from the population via the elitist selection if we add the following rule.

A trial without constraint values will be dominated by trials with constraint values:
1) Return false if neither `trial0` nor `trial1` does not have constraint values
2) Return false if `trial1` has constraint values, but `trial0` does
3) Return true if `trial0` has constraint values, but `trial1` does

This functionality can be reviewed independently, and I'll create a follow-up PR.